File: apps/api/src/services/value_format.py
"""Value formatting on a property (Foundry ``object-link-types`` p.94-101).

> "Value formatting refers to applying a special formatter to the value of a
> property, transforming the raw value to a more readable version." (p.94)

**This module validates a formatter; it does not apply one.** Applying happens
in the browser (``apps/web/src/lib/value-format.ts``), and it has to: p.100
offers "the application user's current timezone" as a legitimate choice, and a
server has no idea what that is. Formatting on the way out would also mean the
API returning `"$100K"` where it used to return `100000`, which would make
every non-screen consumer - filters, actions, aggregations, exports - wrong at
the same time.

What the server owes is a *refusal*, and the refusals are the point of the
file. Every one of them is a page that would otherwise render wrongly or throw:

* a formatter that does not match the property's base type (p.95: "you will see
  a type of formatting depending on the base type of the property"), which
  would be a number formatter quietly doing nothing to a string forever;
* digit bounds `Intl.NumberFormat` throws a `RangeError` on, which in a browser
  is a blank component and a console message nobody reads;
* a style missing the thing that style is *for* - a currency with no currency
  code, a unit with no unit.

The alternative to refusing here is discovering all three by looking at a
screen, which is exactly the trade `STATUS.md` keeps making the other way.
"""
from __future__ import annotations

from typing import Any

# The base types each family of formatter may be attached to (p.95).
NUMERIC_TYPES = ("integer", "float")
TEMPORAL_TYPES = ("date", "timestamp")

KINDS = ("number", "datetime")

#: p.97's "Base type" dropdown, minus Fixed Values - a value-to-label map is a
#: lookup rather than a formatter, and it is named in the parity doc instead of
#: being half-built here.
NUMBER_STYLES = ("plain", "currency", "unit", "percent", "affix")

#: p.99's table, one style per row.
DATETIME_STYLES = ("date", "datetime_long", "datetime_short", "iso", "relative", "time")

#: `Intl.NumberFormat`'s own ranges. Outside them it throws, and a formatter
#: that throws is a component that does not render.
DIGIT_BOUNDS = {
    "minimum_integer_digits": (1, 21),
    "minimum_fraction_digits": (0, 100),
    "maximum_fraction_digits": (0, 100),
    "minimum_significant_digits": (1, 21),
    "maximum_significant_digits": (1, 21),
}

NOTATIONS = ("standard", "compact", "scientific", "engineering")

_NUMBER_FIELDS = (
    "style", "currency", "unit", "prefix", "suffix", "grouping", "notation",
    *DIGIT_BOUNDS,
)
_DATETIME_FIELDS = ("style", "timezone")


class FormatError(ValueError):
    """A formatter that would not render. Surfaced as a 422 with its message."""


def parse(raw: Any, *, data_type: str, property_name: str) -> dict[str, Any] | None:
    """Validate one property's ``value_format``, or refuse it by name.

    Returns the normalised formatter, or ``None`` for "no formatting" - which
    is what every property that has never been configured means, and what
    clearing one means too.

    `property_name` is in every message on purpose: a type is saved with all of
    its properties at once, so "value_format: unknown style" would leave
    somebody hunting through fifteen of them for the one that was wrong.
    """
    if raw is None:
        return None
    if not isinstance(raw, dict):
        raise FormatError(f"{property_name}: value_format must be an object")

    kind = raw.get("kind")
    if kind not in KINDS:
        raise FormatError(
            f"{property_name}: value_format kind must be one of {', '.join(KINDS)}"
        )
    if kind == "number":
        return _number(raw, data_type=data_type, property_name=property_name)
    return _datetime(raw, data_type=data_type, property_name=property_name)


def _unknown_fields(raw: dict[str, Any], allowed: tuple[str, ...]) -> list[str]:
    return sorted(k for k in raw if k != "kind" and k not in allowed)


def _number(
    raw: dict[str, Any], *, data_type: str, property_name: str
) -> dict[str, Any]:
    if data_type not in NUMERIC_TYPES:
        raise FormatError(
            f"{property_name}: number formatting needs a numeric property; "
            f"this one is {data_type}"
        )
    extra = _unknown_fields(raw, _NUMBER_FIELDS)
    if extra:
        # Refused rather than dropped. A misspelled option that is silently
        # ignored is a setting somebody believes they turned on.
        raise FormatError(
            f"{property_name}: unknown number formatting option "
            f"{', '.join(extra)}"
        )

    style = raw.get("style", "plain")
    if style not in NUMBER_STYLES:
        raise FormatError(
            f"{property_name}: number style must be one of "
            f"{', '.join(NUMBER_STYLES)}"
        )
    out: dict[str, Any] = {"kind": "number", "style": style}

    if style == "currency":
        # A three-letter ISO 4217 code, which is what `Intl` takes. Checked for
        # shape rather than against a list: the list changes and this does not
        # need to be the place that tracks it.
        code = raw.get("currency")
        if not isinstance(code, str) or not code.isalpha() or len(code) != 3:
            raise FormatError(
                f"{property_name}: currency formatting needs a three-letter "
                "currency code, like USD"
            )
        out["currency"] = code.upper()
    elif style == "unit":
        unit = raw.get("unit")
        if not isinstance(unit, str) or not unit.strip():
            raise FormatError(
                f"{property_name}: unit formatting needs a unit, like kilogram"
            )
        out["unit"] = unit.strip()
    elif style == "affix":
        prefix = raw.get("prefix") or ""
        suffix = raw.get("suffix") or ""
        for name, value in (("prefix", prefix), ("suffix", suffix)):
            if not isinstance(value, str):
                raise FormatError(f"{property_name}: {name} must be text")
        if not prefix and not suffix:
            # p.97 calls this option "Prefix/Suffix". Neither is not a style,
            # it is `plain` with extra steps, and saving it would make the
            # editor show a style whose effect is nothing.
            raise FormatError(
                f"{property_name}: prefix/suffix formatting needs a prefix or "
                "a suffix"
            )
        out["prefix"], out["suffix"] = prefix, suffix

    if "grouping" in raw:
        if not isinstance(raw["grouping"], bool):
            raise FormatError(f"{property_name}: grouping must be true or false")
        out["grouping"] = raw["grouping"]
    if "notation" in raw:
        if raw["notation"] not in NOTATIONS:
            raise FormatError(
                f"{property_name}: notation must be one of {', '.join(NOTATIONS)}"
            )
        out["notation"] = raw["notation"]

    for field, (low, high) in DIGIT_BOUNDS.items():
        if field not in raw:
            continue
        value = raw[field]
        # `bool` is an `int` in Python and `True` would become 1 digit.
        if not isinstance(value, int) or isinstance(value, bool):
            raise FormatError(f"{property_name}: {field} must be a whole number")
        if not low <= value <= high:
            raise FormatError(
                f"{property_name}: {field} must be between {low} and {high}"
            )
        out[field] = value

    for lo_name, hi_name in (
        ("minimum_fraction_digits", "maximum_fraction_digits"),
        ("minimum_significant_digits", "maximum_significant_digits"),
    ):
        lo, hi = out.get(lo_name), out.get(hi_name)
        if lo is not None and hi is not None and lo > hi:
            # `Intl` throws on this pair, so the component renders nothing at
            # all - the property disappears rather than looking wrong.
            raise FormatError(
                f"{property_name}: {lo_name} cannot be more than {hi_name}"
            )
    return out


def _datetime(
    raw: dict[str, Any], *, data_type: str, property_name: str
) -> dict[str, Any]:
    if data_type not in TEMPORAL_TYPES:
        raise FormatError(
            f"{property_name}: date and time formatting needs a date or "
            f"timestamp property; this one is {data_type}"
        )
    extra = _unknown_fields(raw, _DATETIME_FIELDS)
    if extra:
        raise FormatError(
            f"{property_name}: unknown date and time formatting option "
            f"{', '.join(extra)}"
        )

    style = raw.get("style")
    if style not in DATETIME_STYLES:
        raise FormatError(
            f"{property_name}: date and time style must be one of "
            f"{', '.join(DATETIME_STYLES)}"
        )
    out: dict[str, Any] = {"kind": "datetime", "style": style}

    if "timezone" in raw and raw["timezone"] is not None:
        zone = raw["timezone"]
        if not isinstance(zone, str) or not zone.strip():
            raise FormatError(
                f"{property_name}: timezone must be a zone name, like "
                "Europe/London"
            )
        if data_type != "timestamp":
            # p.100 is specific: "If you are formatting a *timestamp*, you can
            # specify which timezone to render the timestamp". A date has no
            # instant to place in a zone - shifting one by an offset moves it
            # to a different day, which is a wrong answer rather than a
            # differently-presented one.
            raise FormatError(
                f"{property_name}: a timezone applies to a timestamp, not to "
                "a date"
            )
        out["timezone"] = zone.strip()
    # Absent timezone means the viewer's own (p.100), which only the browser
    # knows. Not defaulted to UTC here: a stored "UTC" is an author's choice
    # and an absent one is "wherever the reader is", and collapsing the two
    # would take the choice away.
    return out

File: apps/web/src/lib/value-format.ts
/**
 * Applying a property's value formatter (Foundry `object-link-types`
 * p.94–101).
 *
 * > "Value formatting refers to applying a special formatter to the value of a
 * > property, transforming the raw value to a more readable version … the
 * > weight column [has] a unit ("kg") applied and the value column is
 * > displayed in a more compact form with a currency sign ("$100K")." (p.94)
 *
 * **Here rather than on the server, for a reason the spec states.** p.100
 * offers "the application user's current timezone" as a legitimate choice, and
 * a server does not know what that is. Formatting on the way out would also
 * change what the API returns — `"$100K"` where it used to return `100000` —
 * which makes filters, actions, aggregations and exports wrong all at once.
 * The stored value stays raw; only what a person looks at changes.
 *
 * **A separate rule from conditional formatting**, which colours a value by
 * comparing it. The comparison is against the *raw* value: a rule saying
 * "green above 50000" must not be handed `"$100K"` to compare, because a
 * string never was above anything. This module returns text and nothing else,
 * so the two cannot be confused.
 *
 * Pure on purpose — the same boundary `components/canvas/pure.ts` draws, for
 * the same reason. A formatter is exactly the kind of rule where a browser
 * test proves "something rendered" and a unit test proves *what*.
 */

import type { ValueFormat } from "@/lib/types";

/** Milliseconds in the window p.99 formats relatively. */
const RELATIVE_WINDOW_MS = 24 * 60 * 60 * 1000;

const RELATIVE_UNITS: [Intl.RelativeTimeFormatUnit, number][] = [
  ["hour", 60 * 60 * 1000],
  ["minute", 60 * 1000],
  ["second", 1000],
];

/**
 * The raw value as text, formatted if the property says so.
 *
 * Returns `null` when there is nothing to show, so a caller can tell "no
 * value" from "the empty string" — a formatted zero is `"$0"` and must not be
 * mistaken for a blank.
 *
 * `now` is a parameter rather than a call to `Date.now()` so "8 minutes ago"
 * is testable. A relative format tested against the real clock is a test that
 * asserts the code ran.
 */
export function formatValue(
  raw: unknown,
  format: ValueFormat | null | undefined,
  options: { now?: number; locale?: string } = {},
): string | null {
  if (raw === null || raw === undefined || raw === "") return null;
  if (!format) return String(raw);
  const locale = options.locale ?? "en-US";
  if (format.kind === "number") return formatNumber(raw, format, locale);
  return formatDateTime(raw, format, locale, options.now ?? Date.now());
}

/** p.97–98's options, on `Intl.NumberFormat`.
 *
 * **A value that is not a number comes back as itself.** Properties are stored
 * untyped, so a column declared `float` can hold `"n/a"` from a dataset nobody
 * cleaned. `Intl` renders that as `NaN`, which reads like a computed answer
 * rather than like the text that is actually stored.
 */
function formatNumber(
  raw: unknown,
  format: ValueFormat & { kind: "number" },
  locale: string,
): string {
  const n = typeof raw === "number" ? raw : Number(String(raw).trim());
  if (!Number.isFinite(n)) return String(raw);

  const opts: Intl.NumberFormatOptions = {};
  if (format.style === "currency") {
    opts.style = "currency";
    opts.currency = format.currency;
  } else if (format.style === "unit") {
    opts.style = "unit";
    opts.unit = format.unit;
  } else if (format.style === "percent") {
    opts.style = "percent";
  }
  if (format.grouping !== undefined) opts.useGrouping = format.grouping;
  if (format.notation) opts.notation = format.notation;
  if (format.minimum_integer_digits !== undefined)
    opts.minimumIntegerDigits = format.minimum_integer_digits;
  if (format.minimum_fraction_digits !== undefined)
    opts.minimumFractionDigits = format.minimum_fraction_digits;
  if (format.maximum_fraction_digits !== undefined)
    opts.maximumFractionDigits = format.maximum_fraction_digits;
  if (format.minimum_significant_digits !== undefined)
    opts.minimumSignificantDigits = format.minimum_significant_digits;
  if (format.maximum_significant_digits !== undefined)
    opts.maximumSignificantDigits = format.maximum_significant_digits;

  let text: string;
  try {
    text = new Intl.NumberFormat(locale, opts).format(n);
  } catch {
    // The server refuses the combinations `Intl` throws on, so reaching here
    // means a formatter that predates a rule or arrived some other way. The
    // number is still worth showing — a blank cell is the one outcome that
    // tells a reader nothing.
    text = String(n);
  }
  if (format.style === "affix") {
    return `${format.prefix ?? ""}${text}${format.suffix ?? ""}`;
  }
  return text;
}

/** p.99's table, one branch per row. */
function formatDateTime(
  raw: unknown,
  format: ValueFormat & { kind: "datetime" },
  locale: string,
  now: number,
): string {
  const at = new Date(String(raw));
  // Same reasoning as a non-numeric number: an unparseable date shown as
  // "Invalid Date" hides what is actually stored.
  if (Number.isNaN(at.getTime())) return String(raw);
  const zone = format.timezone;

  if (format.style === "iso") return at.toISOString();
  if (format.style === "relative") return formatRelative(at, locale, now, zone);

  const base: Intl.DateTimeFormatOptions = zone ? { timeZone: zone } : {};
  const styles: Record<string, Intl.DateTimeFormatOptions> = {
    date: { weekday: "short", year: "numeric", month: "short", day: "numeric" },
    datetime_long: {
      weekday: "short", year: "numeric", month: "long", day: "numeric",
      hour: "numeric", minute: "2-digit", second: "2-digit",
    },
    datetime_short: {
      year: "numeric", month: "short", day: "numeric",
      hour: "numeric", minute: "2-digit",
    },
    time: { hour: "numeric", minute: "2-digit" },
  };
  return new Intl.DateTimeFormat(locale, { ...base, ...styles[format.style] }).format(at);
}

/**
 * p.99's footnote, which is a rule rather than a detail:
 *
 * > "When formatting Relative to now, applications will only format in
 * > relative terms up to 24 hours ago. After this, it will render in Date and
 * > time (short) form **with the day of the week**."
 *
 * The weekday is the part worth naming: past the window this is *not* the
 * `datetime_short` style, it is that style plus a weekday, so the two cannot
 * share a branch.
 */
function formatRelative(
  at: Date,
  locale: string,
  now: number,
  zone?: string,
): string {
  const delta = at.getTime() - now;
  if (Math.abs(delta) >= RELATIVE_WINDOW_MS) {
    return new Intl.DateTimeFormat(locale, {
      ...(zone ? { timeZone: zone } : {}),
      weekday: "short", year: "numeric", month: "short", day: "numeric",
      hour: "numeric", minute: "2-digit",
    }).format(at);
  }
  const relative = new Intl.RelativeTimeFormat(locale, { numeric: "auto" });
  for (const [unit, ms] of RELATIVE_UNITS) {
    // Truncated, not rounded. Rounding makes 23h59m read "24 hours ago" — a
    // reading that names the very boundary this branch exists to stay inside,
    // so it looks like the fallback failed. Truncation is also what elapsed
    // time usually means: "8 minutes ago" is *at least* 8.
    if (Math.abs(delta) >= ms) return relative.format(Math.trunc(delta / ms), unit);
  }
  return relative.format(0, "second");
}
